Datatype representation cleanup

pavanbalaji opened the pull request:
This PR contains initial precursor cleanup before adding a new datatype engine.

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
test:jenkins/ch4/ofi

pavanbalaji commented:
test:jenkins/ch4/ofi

pavanbalaji commented:
test:jenkins/ch4/ofi

pavanbalaji commented:
test:jenkins/ch4/ofi

pavanbalaji commented:
test:jenkins/ch4/ofi

pavanbalaji commented:
test:jenkins/ch4/ofi

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

hajimefu commented:
test:intel:ch4/ofi/sockets,psm2/gnu,icc/force-am-disabled,force-am-enabled/auto,netmod,no-odd-even,thread/debug,default

hajimefu commented:
The OFI part of change looks good to me, but let me run the tests internally once more to make sure it doesn't break existing test cases.

pavanbalaji commented:
@minsii and @hajimefu   Can comments?

hajimefu commented:
@pavanbalaji Sorry, still waiting for the test results to come back (our cluster is super crowded this week...). I'll approve once I confirm the results.

minsii commented on src/mpi/datatype/looputil.c:
```diff
@@ -214,8 +214,8 @@ static int external32_float_convert(char *dest_buf,
-                      MPI_Aint count, MPI_Datatype handle, struct MPIR_Segment *segp)
+static int segment_init(const void *buf,
+                        MPI_Aint count, MPI_Datatype handle, struct MPIR_Segment *segp)
 {
```
Can we inline `segment_init ` ? `MPIR_Segment_alloc` may be called at perf-critical path, right ?

pavanbalaji replied:
Done.

minsii commented on src/mpi/datatype/looputil.c:
```diff
@@ -365,12 +365,17 @@ int MPIR_Segment_init(const void *buf,
 
-/* Segment_alloc
+/* MPIR_Segment_alloc
  *
```
This change looks like a pure comment fix, which is unrelated to this commit.

pavanbalaji replied:
Done.

minsii commented on src/mpi/datatype/looputil.c:
```diff
@@ -365,12 +365,17 @@ int MPIR_Segment_init(const void *buf,
-    return (struct MPIR_Segment *) MPL_malloc(sizeof(struct MPIR_Segment), MPL_MEM_DATATYPE);
+    struct MPIR_Segment *segp =
+        (struct MPIR_Segment *) MPL_malloc(sizeof(struct MPIR_Segment), MPL_MEM_DATATYPE);
+
```
We might need to check return of `MPL_malloc` here.

pavanbalaji replied:
Done.

minsii commented on src/mpi/datatype/type_flatten.c:
```diff
@@ -42,12 +41,9 @@ int MPIR_Type_flatten(MPI_Datatype type,
                                                  */
-    if (err)
-        return err;
 
```
We might want to check `segp != NULL` here. I think this fix can be in a different commit.

pavanbalaji replied:
Done.  Btw, this file will be completely rewritten in a follow-up commit that I have coming.

minsii commented on src/mpi/datatype/looputil.c:
```diff
@@ -365,12 +365,17 @@ int MPIR_Segment_init(const void *buf,
+        (struct MPIR_Segment *) MPL_malloc(sizeof(struct MPIR_Segment), MPL_MEM_DATATYPE);
+
+    segment_init(buf, count, handle, segp);
+
```
Can we add a note here explaining that `segment_init ` always return 0, so we do not check the return value and handle the case where `segp` is allocated but `segment_init` fails ?

pavanbalaji replied:
Instead, I've changed the return type of `segment_init` to `void`.

minsii commented on src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c:
```diff
@@ -115,15 +115,15 @@ int MPID_nem_llc_isend(struct MPIDI_VC *vc, const void *buf, int count, MPI_Data
+        struct MPIR_Segment *segment_ptr = MPIR_Segment_alloc(buf, count, datatype);
         MPIR_ERR_CHKANDJUMP(!segment_ptr, mpi_errno, MPI_ERR_OTHER, "**outofmemory");
+
 #ifndef	notdef_leak_0001_hack
```
Please delete the white line change.

minsii commented on src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c:
```diff
@@ -849,7 +847,8 @@ int llc_poll(int in_blocking_poll, llc_send_f sfnc, llc_recv_f rfnc)
-			MPIR_Segment_free(seg);
+
+                        MPIR_Segment_free(seg);
                     }
```
Should here be unchanged ?

pavanbalaji replied:
We are allocating the segment, so it needs to be freed, right?  What's the issue?

minsii replied:
This change is a simple format clean up, right ?

pavanbalaji replied:
Oh, yes.  Sorry, I was looking at the wrong commit.

minsii replied:
This line is still shown as changed. Is it a tab -> space format change ?

pavanbalaji replied:
Fixed.

minsii commented on src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c:
```diff
@@ -1038,13 +1037,13 @@ int MPID_nem_llc_issend(struct MPIDI_VC *vc, const void *buf, int count, MPI_Dat
         /* See MPIDI_CH3_EagerNoncontigSend (in ch3u_eager.c) */
-        struct MPIR_Segment *segment_ptr = MPIR_Segment_alloc();
+        segment_ptr = MPIR_Segment_alloc(buf, count, datatype);
+
```
`segment_ptr ` is not declared.

pavanbalaji replied:
Fixed.

minsii commented on src/mpid/ch3/channels/nemesis/netmod/mxm/mxm_poll.c:
```diff
@@ -447,12 +447,14 @@ static int _mxm_process_rdtype(MPIR_Request ** rreq_p, MPI_Datatype datatype,
+    MPIR_ERR_CHKANDJUMP1((rreq->dev.segment_ptr == NULL), mpi_errno, MPI_ERR_OTHER, "**nomem",
+                         "**nomem %s", "MPIR_Segment_alloc");
+
     rreq->dev.segment_first = 0;
```
Does this change increase overhead ? I.e., every time we have to reallocate buffer for the segment.

pavanbalaji replied:
It does, slightly.  But the bigger cost is in the segment initialization, which we are not avoiding. 
 Also, this is the derived datatype code, which is very slow in the current model.  Once the new derived datatype code is ready, we will not have these alloc/frees in that, which will be much faster.

minsii commented on src/mpid/ch3/channels/nemesis/netmod/mxm/mxm_send.c:
```diff
@@ -299,16 +299,16 @@ int MPID_nem_mxm_send(MPIDI_VC_t * vc, const void *buf, MPI_Aint count, MPI_Data
+                sreq->dev.segment_ptr = MPIR_Segment_alloc(buf, count, datatype);
+                MPIR_ERR_CHKANDJUMP1((sreq->dev.segment_ptr == NULL), mpi_errno, MPI_ERR_OTHER,
+                                     "**nomem", "**nomem %s", "MPIR_Segment_alloc");
                 MPIR_Segment_pack(sreq->dev.segment_ptr, 0, &last, sreq->dev.tmpbuf);
```
The question might be unrelated to this commit: where do we free the segment ?

pavanbalaji replied:
Fixed.

minsii commented on src/mpid/ch3/channels/nemesis/netmod/portals4/ptl_send.c:
```diff
@@ -161,15 +161,15 @@ static int send_msg(ptl_hdr_data_t ssend_flag, struct MPIDI_VC *vc, const void *
         sreq->dev.iov_count = MPL_IOV_LIMIT;
         MPIR_Segment_to_iov(sreq->dev.segment_ptr, sreq->dev.segment_first, &last, sreq->dev.iov, &sreq->dev.iov_count);
+        MPIR_Segment_free(sreq->dev.segment_ptr);
 
```
Looks like you fixed the unreleased segment issue in portals. Could you please also fix the same issue in mxm ? Besides, this fix should be put into a separate commit.

pavanbalaji replied:
Done.

pavanbalaji commented:
test:jenkins/ch3/mxm

minsii commented on src/mpid/ch4/netmod/ofi/ofi_impl.h:
```diff
@@ -707,20 +707,26 @@ MPL_STATIC_INLINE_PREFIX
+        MPIR_Segment_free(dt_seg);
         count1 = params.count;
+
         params.count = 0;
```
White line change

pavanbalaji replied:
Fixed locally.  Will push once the tests complete.

minsii replied:
Is this not pushed ?

pavanbalaji replied:
It is.  The above code shows as "outdated".

minsii commented on src/mpid/ch4/netmod/ofi/ofi_impl.h:
```diff
@@ -707,20 +707,26 @@ MPL_STATIC_INLINE_PREFIX
+        MPIR_Segment_free(dt_seg);
         count2 = params.count;
+
         if (count2 == 1) {      /* Contiguous */
```
While line change

pavanbalaji replied:
Fixed locally.  Will push once the tests complete.

pavanbalaji commented on src/mpi/datatype/dataloop/dataloop_debug.c:
```diff
@@ -0,0 +1,219 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
```
It's not.  It's just moved from a different location.

pavanbalaji commented:
@minsii The PR is ready for your review again.

pavanbalaji commented:
@minsii I think have addressed all of your comments.  Can you take a look at this PR again?

pavanbalaji commented:
Added signoffs and rebased on to the latest `mpich/master`.  Rerunning tests once more before merging.

test:jenkins/ch3/most
test:jenkins/ch4/most
